### PR TITLE
Feature: allow specifying to use before or after other use

### PR DESCRIPTION
**Pull request by eliperelman**

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -750,6 +750,50 @@ config.module
       .options({ presets: ['babel-preset-es2015'] });
 ```
 
+#### Config module rules use before
+
+Specify to that one `use` should be used before another `use`
```
Comment (edmorley) [inline README.md]: Was this meant to be s/to that/that/ ? (And the same below?)

File: src/Rule.js [hunk 2]
```diff
@@ -37,11 +37,29 @@ module.exports = class Rule extends ChainedMap {
...
+    const useEntries = this.uses.entries() || {};
+    const names = Object.keys(useEntries);
+    const uses = [...names];
+
+    names.forEach(name => {
+      const { _before, _after } = useEntries[name];
+
+      if (_before && uses.includes(_before)) {
+        uses.splice(uses.indexOf(name), 1);
+        uses.splice(Math.max(uses.indexOf(_before), 0), 0, name);
```
Comment (edmorley) [inline src/Rule.js]: Why is the `Math.max()` needed? presumably the only time it's required is if `uses.indexOf(_before) < 0` (eg -1 for not found), however the `uses.includes(_before)` presumably rules out that case?

File: src/Rule.js [hunk 3]
```diff
@@ -37,11 +37,29 @@ module.exports = class Rule extends ChainedMap {
   }
 
   toConfig() {
+    const useEntries = this.uses.entries() || {};
+    const names = Object.keys(useEntries);
+    const uses = [...names];
+
+    names.forEach(name => {
+      const { _before, _after } = useEntries[name];
+
```
Comment (edmorley) [inline src/Rule.js]: Do we need a check to ensure both `_before` and `_after` are not set for the same rule, and fail loudly if so? (And a mention in the docs?) Since currently if both are set, the `_after` clobbers the splices made by the `_before` handling.

File: README.md [hunk 4]
```diff
@@ -750,6 +750,50 @@ config.module
       .options({ presets: ['babel-preset-es2015'] });
 ```
 
+#### Config module rules use before
```
Comment (edmorley) [inline README.md]: The other sections use a naming of form "Config module rules uses (loaders): modifying options" - I don't have a preference as to which to use, but wanted to note that this wasn't consistent with them, in case this wasn't intentional.

File: src/Rule.js (hunk 2 above)
Comment (eliperelman) [reply src/Rule.js]: My mistake, this was leftover from a different implementation version.

**Comment by eliperelman** [comment]
@edmorley Great feedback! In order to make ordering more generic for usage with more than `use`s, I changed quite a bit of this PR. All the tests should still pass, but now the `before` and `after` methods can be applied anywhere we like, and are mergeable as well. Also merging should be smarter and less verbose to perform internally after this.

File: README.md [hunk 5]
```diff
@@ -183,34 +183,47 @@ has(key)
...
+// This will order keys if before() or after() are used.
 // returns: Object, undefined if empty
 entries()
 ````
 
 ```js
 // Provide an object which maps its properties and values 
-// into the backing Map as keys and values.
+// into the backing Map as keys and values. Can specify
+// an array of keys to omit from the merge.
```
Comment (edmorley) [inline README.md]: When I read this I was imagining the keys being omitted might be those on the original object (eg so a new array could be passed in for a particular property and not just merged in with the existing array). However from reading the source below the opposite appears to be the case? (I'm struggling at quick glance to tell though due to the larger diff and not having spent long looking at it) Either way I think it would be good to clarify in this comment :-)

File: README.md [hunk 6]
```diff
@@ -600,6 +613,48 @@ config
   .init((Plugin, args) => new Plugin(...args));
 ```
 
+#### Config plugins: ordering before
+
+Specify that the current `plugin` context should operate before another named `plugin`.
+You cannot use `before()` and `after()` on the same plugin.
```
Comment (edmorley) [inline README.md]: s/cannot use/cannot use both/ perhaps? (And in the other instances below)

File: README.md [hunk 7]
```diff
@@ -600,6 +613,48 @@ config
...
+```js
+config
+  .plugin(name)
+    .before(otherName)
+    
+// Example
+
+config
+  .plugin('html-template')
+    .use(HtmlWebpackTemplate)
```
Comment (edmorley) [inline README.md]: The indentation here is inconsistent with some of the other existing entries - this indents the `.use()` one level deeper than `.plugin()`, the others don't. Thoughts on which is preferred?

File: src/ChainedMap.js [hunk 8]
```diff
@@ -82,13 +118,43 @@ module.exports = class extends Chainable {
...
+      throw new Error(`Unable to set before(${JSON.stringify(name)}) with existing value for after`);
+    }
+
+    this.__before = name;
+    return this;
+  }
+
+  after(name) {
+    if (this.__after) {
+      throw new Error(`Unable to set after(${JSON.stringify(name)}) with existing value for before`);
```
Comment (edmorley) [inline src/ChainedMap.js]: Should this be `after(...)` or `.after(...)` ? (And same for `before(...)`)

I can't decide :-)

File: src/ChainedMap.js [hunk 9]
```diff
@@ -82,13 +118,43 @@ module.exports = class extends Chainable {
...
+    }
+
+    this.__after = name;
+    return this;
+  }
+
+  merge(obj, omit = []) {
+    if (obj.before && obj.after) {
+      throw new Error(`Unable to merge before: ${JSON.stringify(obj.before)} and after: ${JSON.stringify(obj.after)} on the same object`);
```
Comment (edmorley) [inline src/ChainedMap.js]: Can this case be hit given the checks at time of `before()` and `after()` use?

File: src/ChainedSet.js [hunk 10]
```diff
@@ -39,11 +39,11 @@ module.exports = class extends Chainable {
     return this;
   }
 
-  when(condition, trueBrancher = Function.prototype, falseBrancher = Function.prototype) {
+  when(condition, whenTruthy = Function.prototype, whenFalsy = Function.prototype) {
```
Comment (edmorley) [inline src/ChainedSet.js]: This could be a separate commit in the PR :-)

File: src/ChainedMap.js [hunk 11]
```diff
@@ -53,8 +74,23 @@ module.exports = class extends Chainable {
     return this;
   }
 
-  merge(obj) {
-    Object.keys(obj).forEach(key => this.set(key, obj[key]));
+  merge(obj, omit = []) {
```
Comment (edmorley) [inline src/ChainedMap.js]: This and the other merge/omit parts could be a separate commit in the PR from the before/after parts :-)

File: src/Plugin.js [hunk 12]
```diff
@@ -28,12 +28,12 @@ module.exports = class extends ChainedMap {
       this.set('args', obj.args);
     }
 
-    return this;
+    return super.merge(obj, [...omit, 'args', 'plugin'])
```
Comment (edmorley) [inline src/Plugin.js]: Missing semicolon.

I wonder if we should add some linting to this repo?

File: test/Rule.js [hunk 13]
```diff
@@ -113,6 +113,106 @@ test('toConfig with values', t => {
...
+      {
+        loader: 'babel-loader',
+        options: {
+          presets: ['alpha']
+        }
+      }
+    ]
+  });
+});
+
```
Comment (edmorley) [inline test/Rule.js]: Could you add a test for the "before and after aren't allowed at the same time" handling?

I think we also need some tests for all of the merge omit functionality?

File: README.md (hunk 5 above)
Comment (eliperelman) [reply README.md]: Yeah, I'll try to phrase this better. The `omit` property is supposed to be an array of keys that should be excluded from merging into the config, e.g. the following would do nothing.

```js
config.merge({ devtool: 'source-map' }, ['devtool'])
```

File: src/ChainedMap.js (hunk 9 above)
Comment (eliperelman) [reply src/ChainedMap.js]: Yes, when merging in configuration from an object, these can be set by the user:

```js
config.merge({
  module: {
    rule: {
      compile: {
        use: {
          cache: {
            loader: 'cache-loader',
            before: 'babel'
          }
        }
      }
    }
  }
})
```

File: src/Plugin.js (hunk 12 above)
Comment (eliperelman) [reply src/Plugin.js]: Yeah, I've put it off for a while. I can do that in another PR, although it would be easier to either use Neutrino (seems circular) or move this repo into the monorepo.

**Comment by eliperelman** [comment]
OK, I **think** I got everything addressed. Thank you for the in-depth review!

File: README.md [hunk 14]
```diff
@@ -201,8 +201,11 @@ entries()
 ```js
 // Provide an object which maps its properties and values
 // into the backing Map as keys and values.
+// You can also provide an array to the second argument
```
Comment (edmorley) [inline README.md]: s/an array to the second argument/an array as the second argument/ maybe? Up to you :-)

File: README.md [hunk 15]
```diff
@@ -650,6 +693,47 @@ config.resolve
...
+  .end()
+.plugin('alpha')
+  .use(AlphaWebpackPlugin)
+  .before('beta');
+```
+
+#### Config resolve plugins: ordering after
+
+Specify that the current `plugin` context should operate after another named `plugin`.
+You cannot use `.before()` and `.after()` on the same resolve plugin.
```
Comment (edmorley) [inline README.md]: s/cannot use/cannot use both/

File: src/Rule.js [hunk 16]
```diff
@@ -40,45 +40,36 @@ module.exports = class Rule extends ChainedMap {
...
+        .forEach(name => this.use(name).merge(obj.use[name]));
+    }
 
-          case 'test': {
-            return this.test(value instanceof RegExp ? value : new RegExp(value));
-          }
+    if (!omit.includes('oneOf') && 'oneOf' in obj) {
+      Object
+        .keys(obj.oneOf)
+        .forEach(name => this.oneOf(name).merge(obj.oneOf[name]))
```
Comment (edmorley) [inline src/Rule.js]: Missing semicolon

File: README.md [hunk 17]
```diff
@@ -194,6 +193,8 @@ values()
 // where the key is the object property, and the value
 // corresponding to the key. Will return `undefined` if the backing
 // Map is empty.
+// This will order properties by their name if the value is also
```
Comment (edmorley) [inline README.md]: s/also// ? Though happy for you to pick

File: README.md [hunk 18]
```diff
@@ -650,6 +693,47 @@ config.resolve
...
+You cannot use both `.before()` and `.after()` on the same resolve plugin.
+
+```js
+config.resolve
+  .plugin(name)
+    .before(otherName)
+
+// Example
+
+config.resolve.plugin('beta')
```
Comment (edmorley) [inline README.md]: The non-resolve examples and also the config.resolve "after" example below, put the `.plugin()` on a new line, so I think this should be too?

File: src/Orderable.js [hunk 19]
```diff
@@ -0,0 +1,31 @@
+module.exports = (Class) => class extends Class {
+  before(name) {
+    if (this.__after) {
+      throw new Error(`Unable to set .before(${JSON.stringify(name)}) with existing value for after`);
```
Comment (edmorley) [inline src/Orderable.js]: Does the `with existing value for after` part need quotes or something around the "after", to make it clearer? (Or else `.after()` ?)

(And same for below)

**Comment by eliperelman** [comment]
Accepted all your requests.

**Comment by eliperelman** [comment]
🎉
